### ISSUE TITLE
fatal error: color/color.hpp: No such file or directory

### ISSUE THREAD

**Issue by erreubequis** [jpcima/pd-visualization#1]
hi jpcima,
I'm trying to compile your object for pd-vanila on a Raspian stretch Patchbox O.S Real time, and i get the following error due to a missing library:

$ make
[ 19%] Built target visu_common
[ 28%] Built target sgvisu_tilde
[ 38%] Built target ogvisu_tilde
[ 47%] Built target Fl_Knob
[ 57%] Built target wfvisu_tilde
Scanning dependencies of target visu_gui
[ 61%] Building CXX object CMakeFiles/visu_gui.dir/src/gui/w_dft_spectrogram.cc.o
/home/patch/spectros/pd-visualization/src/gui/w_dft_spectrogram.cc:8:27: fatal error: color/color.hpp: No such file or directory
 #include <color/color.hpp>
                           ^
compilation terminated.
CMakeFiles/visu_gui.dir/build.make:86: recipe for target 'CMakeFiles/visu_gui.dir/src/gui/w_dft_spectrogram.cc.o' failed
make[2]: *** [CMakeFiles/visu_gui.dir/src/gui/w_dft_spectrogram.cc.o] Error 1
CMakeFiles/Makefile2:252: recipe for target 'CMakeFiles/visu_gui.dir/all' failed
make[1]: *** [CMakeFiles/visu_gui.dir/all] Error 2
Makefile:127: recipe for target 'all' failed
make: *** [all] Error 2

any clue where I could find this color.hpp file?
thanks in advance

**Comment by jpcima** [comment]
Hi @erreubequis  The color package is available as a submodule repository.
This command should be able to fix your build : `git submodule update --init`

**Comment by erreubequis** [comment]
Hi! The build was inmediatelly succesful with your command, worked straight forward.
The 3 objects are working very goood and the response are great under a RP3+ on a RT kernel.
I'm wondering if there is a way to visualice the objects into the patch and not on a separate window. 
Merci.

**Comment by jpcima** [comment]
No these visualizers are designed to operate as external windows, and it is a feature.
They can't be easily rewritten as patch UI elements.

It's the same case as GEM, running as a separate window.
As far as I'm aware, it's not possible to embed a native window into the canvas.

As oscilloscope and spectrum as concerned, you could certainly write some equivalent displays using Pd directly. IIRC, I think one is documented in the Pd docs.

**Comment by erreubequis** [comment]
Ok I understand. 
I´m very much interested in the waterfall to work on some SDR project. 
Right now I can work prety good with it, and start to adapt to the needs of the project.
The main issue I see with the windows its that they open on _random_ places each time I open the patch for the first time in the session and many times overlapped. Some short of coordenades / window dimension to be understand by X11.
thanks for the fast answers!

**Comment by jpcima** [comment]
Ok. dou you need special ability to set window geometry and position from the pd patch?

**Comment by erreubequis** [comment]
yes, this will be ideal so everytime the patch gets started remembers that positions.
For the last ideal setup,I wont be using the Windows Manager (xfce) as I´m starting the pd patch before the windows manager starts

**Comment by jpcima** [comment]
I've implemented an ability to do such setup.

It works as indicated on the picture:
![Capture du 2019-07-26 21-35-14](https://user-images.githubusercontent.com/17614485/61976934-5c2be000-afed-11e9-8ec1-cfde9d9fa3d2.png)

1. when the window is ready to accepting configuration, the 1st output is banged (DSP must run)
2. in reaction, following messages can be sent:
  - position X Y, set window position
  - size X Y, set window size
  - border B, enable/disable the window frame (1:on 0:off)

This ensures the synchronization such that messages are sent in good timing. Otherwise, it can be ignored; for example, during the loadbang time, if messages go when it's not yet finished starting.

**Comment by erreubequis** [comment]
Hallo Jpcima, that seems perfekt.
As the default wfvisu~ object does not have an output, I've tested different delay times, but no luck yet.
Are you planing to make the commit for this upgrade? 
thanks!

**Comment by jpcima** [comment]
Apologies that I made a mistake here.
I had pushed only at the mirror due to configuration error.
Now, it must be synchronized. (at the head commit b303c45)

(the mirror is located at bitbucket: https://bitbucket.org/jpcima/pd-visualization/src/master/)

**Comment by erreubequis** [comment]
Hi! no worries, i was trying to guess the issue.
I got this problem after the pull
pd-visualization/build $ make
[ 62%] Built target visu_common
[ 62%] Built target sgvisu_tilde
[ 62%] Built target ogvisu_tilde
[ 62%] Built target Fl_Knob
[ 62%] Built target wfvisu_tilde
[ 62%] Built target tfvisu_tilde
[ 62%] Building CXX object CMakeFiles/visu_gui.dir/src/visu~-gui.cc.o
/home/patch/spectros/pd-visualization/src/visu~-gui.cc:13:20: fatal error: fftw3.h: No such file or directory 
#include <fftw3.h> 
^
compilation terminated.
CMakeFiles/visu_gui.dir/build.make:62: recipe for target 'CMakeFiles/visu_gui.dir/src/visu~-gui.cc.o' failed
make[2]: *** [CMakeFiles/visu_gui.dir/src/visu~-gui.cc.o] Error 1
CMakeFiles/Makefile2:289: recipe for target 'CMakeFiles/visu_gui.dir/all' failed
make[1]: *** [CMakeFiles/visu_gui.dir/all] Error 2
Makefile:127: recipe for target 'all' failed
make: *** [all] Error 2
getting close!

**Comment by jpcima** [comment]
It's in package libfftw3-dev.
After some time, I've decided to drop the kft FFT in favor of FFTW, considering FFTW has become faster and KFR devs did not answer issues.

**Comment by erreubequis** [comment]
hi again jpcima,
i'm getting closer:
$ make
[ 62%] Built target visu_common
[ 62%] Built target sgvisu_tilde
[ 62%] Built target ogvisu_tilde
[ 62%] Built target Fl_Knob
[ 62%] Built target wfvisu_tilde
[ 62%] Built target tfvisu_tilde
make[2]: *** No rule to make target '/usr/lib/arm-linux-gnueabihf/libSM.so', needed by '../visu~-gui'.  Stop.
CMakeFiles/Makefile2:289: recipe for target 'CMakeFiles/visu_gui.dir/all' failed
make[1]: *** [CMakeFiles/visu_gui.dir/all] Error 2
Makefile:127: recipe for target 'all' failed
make: *** [all] Error 2

Any clue about this? I'm checking on different forums but not get a solution.

**Comment by jpcima** [comment]
This is libsm-dev. Most likely, it's pulled from FLTK, I don't use this library directly.
You can determine this by debian's package search engine located at https://packages.debian.org/. (search in package content)

**Comment by erreubequis** [comment]
great thanks| i didnt search there 
A few installs and makes after I reached to 
$ make
[ 62%] Built target visu_common
[ 62%] Built target sgvisu_tilde
[ 62%] Built target ogvisu_tilde
[ 62%] Built target Fl_Knob
[ 62%] Built target wfvisu_tilde
[ 62%] Built target tfvisu_tilde
[ 62%] Linking CXX executable "../visu~-gui"
[ 62%] Built target visu_gui
Scanning dependencies of target visu_jack
[ 62%] Building CXX object CMakeFiles/visu_jack.dir/src/jack-main.cc.o
/home/patch/spectros/pd-visualization/src/jack-main.cc:5:23: fatal error: jack/jack.h: No such file or directory
 #include <jack/jack.h>
                       ^
compilation terminated.
CMakeFiles/visu_jack.dir/build.make:62: recipe for target 'CMakeFiles/visu_jack.dir/src/jack-main.cc.o' failed
make[2]: *** [CMakeFiles/visu_jack.dir/src/jack-main.cc.o] Error 1
CMakeFiles/Makefile2:326: recipe for target 'CMakeFiles/visu_jack.dir/all' failed
make[1]: *** [CMakeFiles/visu_jack.dir/all] Error 2
Makefile:127: recipe for target 'all' failed
make: *** [all] Error 2

**Comment by jpcima** [comment]
libjack-dev or libjack-jackd2-dev, depends which version is used

**Comment by erreubequis** [comment]
Yes! 
Make succesfully done and now windows respond to coodenades and can be set perfect.
As you read from my novice writing and doubts this is prety new to me, but thanks to your help and support now its working!
I would like to dig into the color schemes, lets see if i dont break it.
Muchas gracias Jpcima!
